setup.py

```diff
@@ -1,19 +1,6 @@
-from setuptools import setup, find_packages
-from diffkemp.utils import get_simpll_build_dir
-import os
-import subprocess
-
-setup(name="diffkemp",
-      version="0.6.0",
-      description="A tool for semantic difference analysis of kernel functions",
-      author="Viktor Malik",
-      author_email="vmalik@redhat.com",
-      url="https://github.com/viktormalik/diffkemp",
-      packages=find_packages(),
-      setup_requires=["cffi"],
-      install_requires=["pyyaml", "cffi"],
-      cffi_modules=["./diffkemp/simpll/simpll_build.py:ffibuilder"])
-
-if os.path.isfile("diffkemp/simpll/_simpll.abi3.so"):
-    build_dir = get_simpll_build_dir()
-    subprocess.run(["cp", "diffkemp/simpll/_simpll.abi3.so", build_dir])
+from setuptools import setup
```

Comment:
We could probably add here
```python
import os                                                                       
os.environ["SIMPLL_BUILD_DIR"] = "build"
```
to get rid of the `diffkemp` import in `simpll_build.py` file which causes problems in local builds.

Reply:
This could work, but I am not sure whether this enviromental variable would persist because of some deep `pip` sheningans.

I would also be really happy to get rid of `setup.py` altogether, but unfortunately, CFFI developers refuse to implement `pyproject.toml` plugin until they absolutely have to. So we should probably search for a different solution.

It is true that we could perhaps do some simpler implementation of the `get_simpll_build_dir` function, but I have to think about that.

Otherwise this would be a good solution.

Reply:
+1 to getting rid of `setup.py`.

I'm not very familiar with the CFFI build process but couldn't we fully run it manually from CMake? What is `setup.py` doing at the moment?

Reply:
We are calling the manual build from CMake, but that is working only for the "development build". Unfortunately, for installation through `pip` the `setup.py` is hard requirement at the moment and it can only be changed if we use different backend for bindings or if the developers of CFFI library create their own plugin.

For completeness, the code in `setup.py` that we still have is telling the `cffi` library where to look for code from which we want to create bindings.

Reply:
I think (if I understand it correctly), that probably the most important thing that the `setup.py` file does is when running `pip install .`, it creates the `_simpll.xxxx.so` file (extension module which can be imported from Python and used for running the SimpLL library) and **saves it to the `site-packages` directory** which enables that diffkemp/Python is enable to import the module (respectively that it will find the module). So I think the biggest problem which would have to be done in the `CMakefile` file (to get rid of the `setup.py` file) is to copy the `_simpll.xxxx.so` file manually somewhere from where `diffkemp`/Python could import it. I don't know if that could be done or also if it is a good idea to do it manually.

Reply:
Right, so, if I get the process right (please correct me if I'm wrong), the sequence is:
1. build `simpll-lib`
2. run `simpll_build.py` which creates `_simpll.c` (which somehow includes `simpll-lib`)
3. build `_simpll.c` into `_simpll.xxx.so` (extension module)
4. store `simpll.xxx.so` somewhere Python can find and import it (preferably `site-packages`)

IIUC, step 1 is always done in CMake. Steps 2-4 are done by `pip install .`, thanks to `setup.py`. We can also run step 2 in the CMake but I don't see a way to do steps 3 and 4 there. But what confuses me, then, is how the development build can work without steps 3 and 4 (and therefore without `pip install .`).

Reply:
I think that the `_simpll.xxx.so` file is created also by the `simpll_build.py` script. The location of  `_simpll.xxx.so` file is dependent on from where the `simpll_build.py` script is run and I think the current location of `_simpll.xxx.so` is like this:
- if it is run from `setup.py` by `pip install .` it is placed in `xxxx/site-packages`,
- if it is run from `setup.py` by `pip install -e .` it is placed in the directory with diffkemp repo,
- if it is run by [`cmake`](https://github.com/DanielKriz/diffkemp/blob/ef60371853ead2f8218414c953335336099a6b6d/diffkemp/simpll/CMakeLists.txt#L55) it is placed in `BUILDIR/diffkemp/simpll` directory.

I think that the development build works because in the `build/bin/diffkemp` file (respectively in its [template](https://github.com/DanielKriz/diffkemp/blob/ef60371853ead2f8218414c953335336099a6b6d/diffkemp/diffkemp.in#L5)) is added path to `BUILDIR/diffkemp/simpll` directory to `sys.path`. When module is imported in Python, [Python searches for modules in directories located in `sys.path`](https://docs.python.org/3/tutorial/modules.html#the-module-search-path), therefore it is possible to [import the `_simpll` module](https://github.com/DanielKriz/diffkemp/blob/ef60371853ead2f8218414c953335336099a6b6d/diffkemp/simpll/simpll_lib.py#L6) and run the SimpLL library using CFFI.

The problem with release build would be probably that we would have to save the `_simpll.xxx.so` file somewhere else than the `BUILDIR/diffkemp/simpll` directory and then add to `sys.path` the path to it -- it could maybe be done but it would be more complex (I think).

@DanielKriz feel free to correct me.

Reply:
Right, this makes much more sense now, thanks @PLukas2018!

So from the above list, the only extra part handled by `setup.py` and not by CMake is point 4. I agree that it probably doesn't make sense to try to place the `.so` file to `site-packages` (or update `sys.path` in release builds) by our own as we would very likely mess things up.

But, since there seem to be problems with doing local CMake build + `pip install .`, I'm thinking about not supporting such an option at all. In the end, it doesn't make much sense - you want to run part of DiffKemp (SimpLL) from a local build but another part (Python) from installation. I'd say that you should either:
- run everything from a local build, i.e. not do `pip install .` and use the `.so` file from the build directory by using the generated `build/bin/diffkemp`,
- or run everything from an installation, i.e. run `pip install .` but in that case you'd need to do `ninja install`, too, and run the installed `diffkemp`.

This is probably not a best practice as one of the advantages of virtualenv is that it allows you to locally install the developed package without messing up the rest of your system but I'd argue that our setup (Python + CFFI + C++) isn't standard at all and not all best practices apply here.

Reply:
It could be done in CMake, `virtualenv` is setting `VIRTUAL_ENV` variable from which we could read and then install the bindings to correct location, but then we would basically do what `pip` is doing but in non-standard way, which I think is generally bad practice.

It is true, that combination of CFFI and C++ is non-standard, but I think that a better way of handling this issue would be to keep the `setup.py` at minimum (which is the code for invoking the CFFI) and then find more C++ friendly alternative, which could be e.g. `pybind11`.

After this PR and the few small updates to the code base that I have separated from this PR I could look into it. From what I have read about `pybind` it should be (hopefully) drop in replacement for us and it has CMake support.

- - -
To give you complete example, the `cffi` module is actually calling the build script with some arguments, when I was prototyping I did try to write `raise RuntimeError(sys.argv)` at the start of the build script and the `cffi` is passing some arguments to it, but I didn't find them significant.

Reply:
> It could be done in CMake, `virtualenv` is setting `VIRTUAL_ENV` variable from which we could read and then install the bindings to correct location, but then we would basically do what `pip` is doing but in non-standard way, which I think is generally bad practice.

Yeah, agreed.

> It is true, that combination of CFFI and C++ is non-standard, but I think that a better way of handling this issue would be to keep the `setup.py` at minimum (which is the code for invoking the CFFI) and then find more C++ friendly alternative, which could be e.g. `pybind11`.

That's my recommendation, too. I was just under impression that some discussion here led to us supporting local CMake build + `pip install .` which I don't think is a good approach (for reasons explained above). I think that omitting that option will let keep `setup.py` (as well as other code) simpler.

> After this PR and the few small updates to the code base that I have separated from this PR I could look into it. From what I have read about `pybind` it should be (hopefully) drop in replacement for us and it has CMake support.

That would be great!

> To give you a complete example, the `cffi` module is actually calling the build script with some arguments, when I was prototyping I did try to write `raise RuntimeError(sys.argv)` at the start of the build script and the `cffi` is passing some arguments to it, but I didn't find them significant.

Thanks, that's useful. As long as it works, I believe that we're fine (but feel free to post the arguments here for double-checking if you feel like it).